src/common/src/array/column_proto_readers.rs

```diff
@@ -68,7 +68,12 @@ pub fn read_bool_array(array: &ProstArray, cardinality: usize) -> ArrayResult<Ar
     let bitmap: Bitmap = array.get_null_bitmap()?.into();
 
     let arr = BoolArray::new(data, bitmap);
-    assert_eq!(arr.len(), cardinality);
+    // Either array length should match cardinality,
+    // or it is null, and array length should be 0.
+    // To avoid confusion, "len" here does not refer to length of array,
+    // rather, it is the number of rows of arrays returned.
+    // cardinality also refers to this.
+    assert!(arr.len() == cardinality || arr.len() == 0);
```

Comment:
cc @xiangjinwu is there a better assertion we can make for `null` case?

`arr.len() == 0` doesn't seem very precise...

Reply:
Added a e2e test case for this. Previously it panics on this assertion, see e2e test case for concrete sql.

Reply:
This looks like a bug in the caller or even sender encoding side. I am taking a look.

Reply:
#8082 done